### PR TITLE
Change sgx-sdk to sgx-ssl for Cryptographic Functionalities APIs

### PR DESCRIPTION
**Pull request by wanghouqi**
these restful API will be adjusted:

POST <ehsm_srv_address>/ehsm?Action=
        CreateKey
        Encrypt
        Decrypt
        AsymmetricEncrypt
        AsymmetricDecrypt
        Sign
        Verify
        GenerateDataKey
        GenerateDataKeyWithoutPlaintext
        ExportDataKey

test: passed the unittest

Signed-off-by: wanghouqi <wanghouqi@vip.qq.com>

### CODE REVIEW

File: include/datatypes.h [hunk 1]
```diff
@@ -38,69 +38,78 @@
...
 #define DATATYPES_H_
 
-#define DH_KEY_SIZE        20
+#define RSA_PKCS1_PADDING       1
+#define RSA_SSLV23_PADDING      2
+#define RSA_NO_PADDING          3
+#define RSA_PKCS1_OAEP_PADDING  4
+#define RSA_X931_PADDING        5
+/* EVP_PKEY_ only */
+#define RSA_PKCS1_PSS_PADDING   6
```
Comment (syan10) [inline include/datatypes.h:47]: what does it mean? if they are just flags, suggest to put them into a enum{}

File: include/datatypes.h [hunk 2]
```diff
@@ -134,7 +143,93 @@ typedef struct _session_id_tracker_t
...
+    EH_SHA_2_224,
+    EH_SHA_2_256,
+    EH_SHA_2_384,
+    EH_SHA_2_512,
+    EH_SM3
+} ehsm_digest_mode_t;
+
 #pragma pack(pop)
 
+static uint32_t ehsm_get_rsa_padding_size(uint32_t padding_mode)
```
Comment (syan10) [inline include/datatypes.h:195]: don't put the following function in the header file

File: core/Enclave/enclave_hsm.edl [hunk 3]
```diff
@@ -68,48 +68,67 @@ enclave {
     trusted {
         /* Interfaces for eHSM core crypto functions */
         public sgx_status_t enclave_create_aes_key([out, size=cmk_blob_size] uint8_t *cmk_blob,
-                uint32_t cmk_blob_size, [out] uint32_t *req_blob_size);
+                                                    uint32_t cmk_blob_size,
+                                                    [out] uint32_t *req_blob_size,
+                                                    uint32_t key_size);
+        public sgx_status_t enclave_create_sm4_key([out, size=cmk_blob_size] uint8_t *cmk_blob,
```
Comment (syan10) [inline core/Enclave/enclave_hsm.edl:74]: seems the parameters are same, so we can combine the similar ecalls...

File: core/Enclave/enclave_hsm.edl [hunk 4]
```diff
@@ -68,48 +68,67 @@ enclave {
...
-                                               [out] uint32_t *req_blob_size);
+                                                [in, size=cmk_blob_size] const uint8_t *cmk_blob, size_t cmk_blob_size,
+                                                [in, size=cipherblob_len] const uint8_t *cipherblob, size_t cipherblob_len,
+                                                [out, size=plaintext_len] uint8_t *plaintext, size_t plaintext_len,
+                                                ehsm_keyspec_t keyspec, uint32_t key_size);
+
         public sgx_status_t enclave_rsa_sign([in, size=cmk_blob_size] const uint8_t *cmk_blob, size_t cmk_blob_size,
-                                         [in, size=data_len] const uint8_t *data, uint32_t data_len,
-                                         [out, size=signature_len] uint8_t *signature, uint32_t signature_len);
+                                         uint8_t padding_mode, uint32_t digest_mode, uint8_t keyspec,
```
Comment (syan10) [inline core/Enclave/enclave_hsm.edl:123]: not sure, for example:
enclave_rsa_sign, enclave_ecc_sign, enclave_sm2_sign -> enclave_sign

enclave_aes_encrypt, enclave_sm4_encrypt -> enclave_encrypt
enclave_aes_decrypt, enclave_sm4_decrypt -> enclave_decrypt

File: core/Enclave/enclave_hsm.cpp [hunk 5]
```diff
@@ -111,18 +110,29 @@ void printf(const char *fmt, ...)
     ocall_print_string(buf);
 }
 
-typedef struct _aes_gcm_data_ex_t
+/**
+ * @brief Get the block mode by keyspec
+ * @param keyspec the type of key
+ * @return const CHIPER* (openssl callback, tempoary)
+ */
+const EVP_CIPHER * get_symmetric_block_mode(ehsm_keyspec_t keyspec)
```
Comment (syan10) [inline core/Enclave/enclave_hsm.cpp:118]: if just used as internal, add static

File: core/Enclave/enclave_hsm.cpp [hunk 6]
```diff
@@ -111,18 +110,29 @@ void printf(const char *fmt, ...)
...
+        case EH_SM4:
+            return EVP_sm4_ecb();
+        default:
+            return NULL;
+    }
+    return NULL;
+}
 
-static uint32_t sgx_calc_gcm_data_size(const uint32_t aad_size, const uint32_t plaintext_size)
+static uint32_t ehsm_calc_gcm_data_size(const uint32_t aad_size, const uint32_t plaintext_size)
```
Comment (syan10) [inline core/Enclave/enclave_hsm.cpp:135]: same here, another issue is to remove "ehsm_"

Comment (syan10) [reply core/Enclave/enclave_hsm.cpp:135]: same with other function named as "ehsm_xxx"

File: core/Enclave/enclave_hsm.cpp [hunk 7]
```diff
@@ -147,653 +157,2072 @@ static uint32_t sgx_get_gcm_ciphertext_size(const sgx_aes_gcm_data_ex_t *gcm_dat
...
+*           uint8_t *p_iv - Pointer to initialization vector to use
+*           uint32_t iv_len - Length of initialization vector
+*           uint8_t *p_aad - Pointer to input stream of additional authentication data
+*           uint32_t aad_len - Length of additional authentication data stream
+*           sgx_aes_gcm_128bit_tag_t *p_in_mac - Pointer to expected MAC in decryption process
+*   Output: uint8_t *p_dst - Pointer to cipher text. Size of buffer should be >= src_len.
+*           sgx_aes_gcm_128bit_tag_t *p_out_mac - Pointer to MAC generated from encryption process
+* NOTE: Wrapper is responsible for confirming decryption tag matches encryption tag */
+sgx_status_t aes_gcm_encrypt(const uint8_t *p_key, const uint8_t *p_src, uint32_t src_len,
+                             uint8_t *p_dst, const uint8_t *p_iv, uint32_t iv_len, const uint8_t *p_aad,
```
Comment (syan10) [inline core/Enclave/enclave_hsm.cpp:207]: seems this function not necessary, and many duplicated checks with the caller.

and also this function just called by the enclave_aes_encrypt(), so you can put the logic directly into enclave_aes_encrypt()

File: core/Enclave/enclave_hsm.cpp [hunk 8]
```diff
@@ -147,653 +157,2072 @@ static uint32_t sgx_get_gcm_ciphertext_size(const sgx_aes_gcm_data_ex_t *gcm_dat
...
+    }
     return ret;
 }
 
-sgx_status_t enclave_aes_decrypt(const uint8_t *aad, size_t aad_len,
-                             const uint8_t *cmk_blob, size_t cmk_blob_size,
-                             const uint8_t *cipherblob, size_t cipherblob_len,
-                             uint8_t *plaintext, size_t plaintext_len)
+sgx_status_t aes_gcm_decrypt(const uint8_t *p_key, const uint8_t *p_src,
+                             uint32_t src_len, uint8_t *p_dst, const uint8_t *p_iv, uint32_t iv_len,
```
Comment (syan10) [inline core/Enclave/enclave_hsm.cpp:320]: same with it

File: core/Enclave/enclave_hsm.cpp [hunk 9]
```diff
@@ -147,653 +157,2072 @@ static uint32_t sgx_get_gcm_ciphertext_size(const sgx_aes_gcm_data_ex_t *gcm_dat
...
+ * public key storages with plaintext and private key will be encrypted with domain_key
+ * @param cmk_blob
+ * @param cmk_blob_size
+ * @param req_blob_size
+ * @param keyspec
+ * @param padding_mode
+ * @return sgx_status_t
+ */
+sgx_status_t enclave_create_rsa_key(uint8_t *cmk_blob,
+                                    uint32_t cmk_blob_size,
```
Comment (syan10) [inline core/Enclave/enclave_hsm.cpp:1037]: suggest to combine the symmetric createkey as: enclave_create_key

asymmetric related create key API to: enclave_create_keypair

File: core/Enclave/enclave_hsm.cpp [hunk 10]
```diff
@@ -147,653 +157,2072 @@ static uint32_t sgx_get_gcm_ciphertext_size(const sgx_aes_gcm_data_ex_t *gcm_dat
...
+        int key_len = BIO_pending(bio);
+        if (key_len == 0)
+            break;
+            
+        p_key = (uint8_t*)malloc(key_len + 1); // add '\0'
+
+        p_key[key_len] = '\0';
+
+        BIO_read(bio, (char*)p_key, key_len);
+
```
Comment (syan10) [inline core/Enclave/enclave_hsm.cpp:1130]: seems better than previously.

File: core/Enclave/enclave_hsm.cpp (hunk 5 above)
Comment (houhuayou) [reply core/Enclave/enclave_hsm.cpp:118]: Ok, I'll add static

File: core/Enclave/enclave_hsm.cpp (hunk 7 above)
Comment (houhuayou) [reply core/Enclave/enclave_hsm.cpp:207]: OK, I'll put them into enclave_aes_encrypt() directly.

File: core/Enclave/enclave_hsm.cpp (hunk 8 above)
Comment (houhuayou) [reply core/Enclave/enclave_hsm.cpp:320]: Ok, copy understood

**Comment by wanghouqi** [comment]
https://github.com/intel/ehsm/pull/171

**Pull request closed without merging**